packages/flutter/lib/src/material/expansion_tile.dart

```diff
@@ -576,7 +576,12 @@ class _ExpansionTileState extends State<ExpansionTile> with SingleTickerProvider
     _iconColor = _animationController.drive(_iconColorTween.chain(_easeInTween));
     _backgroundColor = _animationController.drive(_backgroundColorTween.chain(_easeOutTween));
 
-    _isExpanded = PageStorage.maybeOf(context)?.readState(context) as bool? ?? widget.initiallyExpanded;
+    final dynamic pageStore = PageStorage.maybeOf(context)?.readState(context);
+    if (pageStore is bool?) {
```

Comment:
If it is a `bool?`, there's no guarantee that the boolean represents the expanded state, and not some other state stored by another widget. Also, the expanded state boolean might actually being saved higher up in the widget tree, but some descendant `PageStorage` is saving something else as a boolean, meaning that the boolean is being misinterpreted here.  These are all problems that the current `PageStorage` API has before your PR, of course, as discussed in both of the bugs you reference.

It seems to me that this problem will probably need a larger solution: a way to determine the source of the stored data so that we 1) know what it's desired type should be, and 2) know that it came from the right widget and what it represents.

One possible enhancement similar to yours would be to add a generic (<T>) argument to the `readState` function that would then only return objects of that type, and null otherwise.  If no type was specified (i.e. it was `dynamic`), then any type would be allowed, which should keep that change from being breaking.

However, this still doesn't solve the problem of what saved the data, and still means that if a widget saves something of the same type that it could be interpreted incorrectly or, if a different type, that a child could eclipse the desired data. This is only currently solved by adding unique `PageStorageKey`s on things that store state.
 
I haven't looked too closely at solutions for this, but there are probably better ways to solve it than just casting it if it's a bool. Your change avoids the cast error, but it hides other possible errors, and in general we'd rather have a cast error than something silently broken.

Reply:
While I agree that the PageStorage API does need attention in general, I do not understand why PageStorage is managing the expanded state of an ExpansionTile or why the logic prioritizes PageStorage to manage the expanded state over what may be passed in from a developer -- the latter being a fundamental issue. Even if the PageStorage API's problems are resolved, PageStorage being a priority over what a developer passes in to the widget is still a problem.

Reply:
Sorry for the long delay. I missed your response.

We'd still prefer the crash to a `debugPrint`.  It just masks the error, and having it crash makes it clear that there is an error.

The proposal you have here is to ignore the `PageStorage` data if the type is wrong. I think the better solution may be to fix the `ExpansionTile` to add an explicit identifier to the `readState` and `writeState` calls so that something else storing state within the same `PageStorage` scope won't overwrite or conflict with the expansion state.

Reply:
I'll see if I can implement your solution. Unfortunately, it might take me a few weeks to get to, though.